lib/ace/ext/papeeria/tex_completer.coffee

```diff
@@ -86,37 +93,70 @@ define( (require, exports, module) ->
+    return token.type.lastIndexOf(type) > -1
+  
+
+  init = (editor, bindKey) -> 
+    HashHandler = require("ace/keyboard/hash_handler").HashHandler
+    keyboardHandler = new HashHandler()
     keyboardHandler.addCommand(
-      name: 'add item in list mode',
-      bindKey: bindKey,
-      exec: (editor) ->
+      name: 'add item in list mode'
+      bindKey: bindKey
+      exec: (editor) ->  
         pos = editor.getCursorPosition()
-        curLine = editor.session.getLine(pos.row);
+        curLine = editor.session.getLine(pos.row)
+        indentCount = ContextHelper.getNestedListDepth(editor.session, pos.row)
+        tabString = editor.getSession().getTabString()
         # it's temporary fix bug with added \item before \begin{itemize|enumerate}
-        if (editor.getSession().getContext(pos.row) == "list" && curLine.indexOf("begin") < pos.column) 
-          editor.insert("\n\t\\item ")
+        if ContextHelper.getContext(editor.session, pos.row) == LIST_STATE && curLine.indexOf("begin") < pos.column
+          editor.insert("\n" + tabString.repeat(indentCount) + "\\item ")
           return true
         else 
-          return false
+          return false   
     )
     editor.keyBinding.addKeyboardHandler(keyboardHandler)
 
-  getCompletions = (editor, session, pos, prefix, callback) ->
-    context = session.getContext(pos.row)
-    if context == "start"
-      callback(null, listSnippets.concat(equationSnippets.concat(basicSnippets)))
+  class ReferenceGetter
+    constructor: ->
+      @lastFetchedUrl =  ""
+      @cache = []
+    processData: (data) -> 
+      @cache = data.map((elem) => 
+          return {
+            name: elem.caption
+            value: elem.caption
+            meta: elem.type + "-ref" 
+          }
+    )
+    getReferences: (url, callback) => 
+      if url != @lastFetchedUrl
+        $.getJSON(url).done((data) => 
+          @processData(data)
+          callback(null, @cache)
+          @lastFetchedUrl = url
+        )
 
-    if context == LIST_STATE
-      callback(null, listKeywords_.concat(listSnippets.concat(equationSnippets)))
+  class TexCompleter
+      constructor: ->
+        @refGetter = new ReferenceGetter()
+      #callback -- add items in popup
+      getCompletions: (editor, session, pos, prefix, callback) =>
+        context = ContextHelper.getContext(session, pos.row)
+        token = session.getTokenAt(pos.row, pos.column)
 
-    if context == EQUATION_STATE
-      callback(null, formulasSnippets.concat(equationKeywords_))
-  
+        if istype(token, "ref")
+          @refGetter.getReferences("example.json", callback)
+        else if context == "start"
+          callback(null, listSnippets.concat(equationSnippets.concat(basicSnippets)))
+        else if context == LIST_STATE
+          callback(null, listKeywords_.concat(listSnippets.concat(equationSnippets)))
+        else if context == EQUATION_STATE
+          callback(null, formulasSnippets.concat(equationKeywords_))
```

Comment:
can you pls rename `equationKeywords_` to `equationKeywords` and vice versa? This trailing underscore looks annoying.